lib/pbench/test/unit/server/test_schema.py

```diff
@@ -0,0 +1,151 @@
+import dateutil
+import pytest
+from typing import Callable
+
+from pbench.server.api.auth import Auth, UnknownUser
+from pbench.server.api.resources.query_apis import (
+    ParamType,
+    ConversionError,
+    InvalidRequestPayload,
+    MissingParameters,
+    Parameter,
+    Schema,
+)
+
+
+class TestParamType:
+    """
+    Tests on the ParamType enum
+    """
+
+    def test_enum(self):
+        assert len(ParamType.__members__) == 4
+        for n, t in ParamType.__members__.items():
+            assert str(t) == t.friendly.upper()
+            assert isinstance(t.convert, Callable)
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, "x", "x"),
+            (ParamType.JSON, {"key": "value"}, {"key": "value"}),
+            (ParamType.DATE, "2021-06-29", dateutil.parser.parse("2021-06-29")),
+            (ParamType.USER, "drb", "drb"),
+        ),
+    )
+    def test_successful_conversions(self, test, monkeypatch):
+        def ok(user: str) -> str:
+            return user
+
+        monkeypatch.setattr(Auth, "validate_user", ok)
+
+        ptype, value, expected = test
+        result = ptype.convert(value)
+        assert result == expected
+
+    @pytest.mark.parametrize(
+        "test",
+        (
+            (ParamType.STRING, {"not": "string"}),
+            (ParamType.JSON, "not_json"),
```

Comment:
Not to be pedantic, but `"not_json"` is a string which _is_, I believe, valid JSON.  Perhaps `"not_json_object"` would be more accurate and appropriate?

Reply:
Ditto line 141.

Reply:
In terms of our JSON payload expectations and in particular the conversions being tested, a JSON document is a `Dict[AnyStr, Any]` and this is definitely _not_ a `dict`! I suppose I could have written `"not_a_json_dict"` but that seems to just muddy the waters. Python deals with JSON as dictionaries, not as strings.

Reply:
I refer you to [the spec](https://www.json.org/json-en.html).  :-)

What valid JSON _is_ is somewhat separate from how it is represented.  A valid JSON _object_ is commonly represented as a dict (although it could be rendered as a string, using JSON...), but there are other elements which are valid in JSON, including strings.  So, `"not_json"` is, in fact, valid JSON.  :-D

Reply:
> Python deals with JSON as dictionaries, not as strings.

Just to be completist:
```
>>> import json
>>> type(json.loads('"This is JSON"'))
<class 'str'>
>>> type(json.loads('{"object": "JSON"}'))
<class 'dict'>
```
So, Python represents JSON objects as `dict`'s, and JSON strings as `str`'s.